File: Juggler/AudioSession.m
//
//  AudioSession.m
//  Juggler
//
//  Created by Evan Lewis on 6/9/14.
//  Copyright (c) 2014 Evan Lewis. All rights reserved.
//

@import MediaPlayer;
#import "AudioSession.h"

@implementation AudioSession {
}

@synthesize audioPlayer;

- (instancetype)init {
  if (self = [super init]) {
    NSURL* soundFileURL = [[NSBundle mainBundle] URLForResource:@"The_Juggler" withExtension:@"mp3"];
    NSError* error = nil;
    self.audioPlayer = [[AVAudioPlayer alloc]initWithContentsOfURL:soundFileURL error:&error];
    self.audioPlayer.numberOfLoops = -1;
    if (error != nil) {
      DLog(@"Failed to load the sound: %@", [error localizedDescription]);
    }
    [self loadAudioSettings];
  }
  return self;
}

- (void)startAudio {
  if (self.isAudioSessionActive) {
    //[self.audioPlayer prepareToPlay];
    //[self.audioPlayer play];
  }
}

- (void)stopAudio {
  //[self.audioPlayer stop];
  self.isAudioSessionActive = NO;
}

- (void)pauseAudio {
  //[self.audioPlayer pause];
  self.isAudioSessionActive = NO;
}

- (void)resumeAudio {
  if (self.isAudioSessionActive) {
    //[self.audioPlayer play];
  }
  self.isAudioSessionActive = YES;
}

- (SKAction *)playSound:(NSString *)sound {
  if(self.isAudioSessionActive) {
    //SKAction* playSound = [SKAction playSoundFileNamed:sound waitForCompletion:NO];
    //return playSound;
  }
  return nil;
}

- (void)loadAudioSettings {
  
  BOOL isRunMoreThanOnce = [[NSUserDefaults standardUserDefaults] boolForKey:@"isRunMoreThanOnce"];
  if(!isRunMoreThanOnce){
    DLog(@"First Run");
    [[NSUserDefaults standardUserDefaults] setBool:YES forKey:@"isRunMoreThanOnce"];
    [[NSUserDefaults standardUserDefaults] setBool:YES forKey:@"Juggler_isAudioSessionActive"];
    [[NSUserDefaults standardUserDefaults] synchronize];
  }
  BOOL defaults = [[NSUserDefaults standardUserDefaults] boolForKey:@"Juggler_isAudioSessionActive"];
  if ((defaults == YES) && ([[MPMusicPlayerController iPodMusicPlayer] playbackState] != MPMusicPlaybackStatePlaying)) {
    self.isAudioSessionActive = YES;
  } else {
    self.isAudioSessionActive = NO;
  }
  [[NSUserDefaults standardUserDefaults] synchronize];
}

- (BOOL)toggleAudioSettings {
  if (self.isAudioSessionActive) {
    DLog(@"turning off audio");
    //[self stopAudio];
    [[NSUserDefaults standardUserDefaults] setBool:NO forKey:@"Juggler_isAudioSessionActive"];
    [[NSUserDefaults standardUserDefaults] synchronize];
    return NO;
  } else if ([[MPMusicPlayerController iPodMusicPlayer] playbackState] != MPMusicPlaybackStatePlaying) {
    self.isAudioSessionActive = YES;
    //[self startAudio];
    DLog(@"turning on audio");
    [[NSUserDefaults standardUserDefaults] setBool:YES forKey:@"Juggler_isAudioSessionActive"];
    [[NSUserDefaults standardUserDefaults] synchronize];
    return YES;
  }
  return NO;
}


@end

File: Juggler/AudioSession.swift
//
//  AudioSession.swift
//  Juggler
//
//  Created by Evan Lewis on 6/9/16.
//  Copyright © 2016 Evan Lewis. All rights reserved.
//

import Foundation
import MediaPlayer
import SpriteKit

class AudioSession {
  
  var audioPlayer: AVAudioPlayer
  var audioSessionIsActive: Bool // Informs if SpriteKit should play sounds(SpriteKit BUG)
  
  init() {
    audioPlayer = AVAudioPlayer()
    
    if let soundFileURL = NSBundle.mainBundle().URLForResource("The_Juggler", withExtension: "mp3") {
      do {
        try audioPlayer = AVAudioPlayer(contentsOfURL: soundFileURL)
      } catch {
        print(error)
      }
    }
    
    if !NSUserDefaults.standardUserDefaults().boolForKey("isRunMoreThanOnce") {
      print("First Run")
      NSUserDefaults.standardUserDefaults().setBool(true, forKey: "isRunMoreThanOnce")
      NSUserDefaults.standardUserDefaults().setBool(true, forKey: "Juggler_isAudioSessionActive")
    }
    
    if NSUserDefaults.standardUserDefaults().boolForKey("Juggler_isAudioSessionActive") && MPMusicPlayerController.systemMusicPlayer().playbackState != MPMusicPlaybackState.Playing {
      audioSessionIsActive = true
    } else {
      audioSessionIsActive = false
    }
    
    NSUserDefaults.standardUserDefaults().synchronize()
  }
  
  func startAudio() {
    if audioSessionIsActive {
      audioPlayer.prepareToPlay()
      audioPlayer.play()
    }
  }
  
  func stopAudio() {
    audioSessionIsActive = false
  }
  
  func playSound(sound: String) -> SKAction? {
    if self.audioSessionIsActive {
        return SKAction.playSoundFileNamed(sound, waitForCompletion: false)
    }
    return nil
  }
  
  func toggleAudioSettings() -> Bool {
    if (self.audioSessionIsActive) {
      self.stopAudio()
      NSUserDefaults.standardUserDefaults().setBool(false, forKey: "Juggler_isAudioSessionActive")
      NSUserDefaults.standardUserDefaults().synchronize()
      return false;
    } else if MPMusicPlayerController.systemMusicPlayer().playbackState != MPMusicPlaybackState.Playing {
      self.audioSessionIsActive = true;
      self.startAudio()
      
      NSUserDefaults.standardUserDefaults().setBool(true, forKey: "Juggler_isAudioSessionActive")
      NSUserDefaults.standardUserDefaults().synchronize()
      return true;
    }
    return false;
  }
  
}